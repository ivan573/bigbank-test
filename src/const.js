export const Route = {
  HOME: '/',
  GAME: '/game',
  SHOP: '/shop',
  REPUTATION: '/reputation'
}

export const ModalWindowType = {
  QUEST: 'quest',
  QUEST_RESULT: 'questResult',
  PURCHASE_ITEM: 'purchaseItem',
  PURCHASE_RESULT: 'purchaseResult',
  GAME_OVER: 'gameOver'
}
