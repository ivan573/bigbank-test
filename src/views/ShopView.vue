<template>
  <div class="shop">
    <p v-if="posessedItems.length" class="shop__posessed-items">Previously bought: {{ posessedItems.join(', ') }}</p>
    <div class="shop__available-items">
      <h2 class="shop__available-items-title">Available items</h2>
      <GameItems :items="items" class="shop__available-items-list" @itemClick="handleItemClick" />
      <ModalWindow
        v-if="modalWindowType"
        class="shop__modal-window"
        :type="modalWindowType"
        :data="modalWindowData"
        @close="handleModalWindowClose"
        @confirm="handleModalWindowConfirmation"
      />
    </div>
  </div>
</template>

<script>
import { mapState, mapMutations, mapActions } from 'vuex'
import { Route, ModalWindowType } from '@/const'
import GameItems from '@/components/GameItems.vue'
import ModalWindow from '@/components/ModalWindow.vue'

export default {
  name: 'ShopView',
  components: {
    GameItems,
    ModalWindow
  },
  data() {
    return {
      modalWindowType: null
    }
  },
  computed: {
    ...mapState('gameData', {
      gameId: state => state.gameId
    }),
    ...mapState('shop', {
      items: state => state.items,
      activeItem: state => state.activeItem,
      posessedItems: state => state.posessedItems,
      shoppingSuccess: state => state.shoppingSuccess,
      isLoading: state => state.isLoading
    }),
    modalWindowData() {
      switch (this.modalWindowType) {
        case ModalWindowType.PURCHASE_ITEM:
          return this.activeItem
        case ModalWindowType.PURCHASE_RESULT:
          return this.shoppingSuccess
            ? { message: 'You have purchased the item' }
            : { message: 'You cannot purchase this item' }
        default:
          return null
      }
    }
  },
  created() {
    if (!this.gameId) {
      this.$router.push({ path: Route.HOME })
    }
  },
  methods: {
    ...mapMutations('shop', ['updateActiveItem', 'discardActiveItem', 'discardShoppingSuccess']),
    ...mapActions('shop', ['fetchShop', 'purchaseItem']),
    handleItemClick(id) {
      this.updateActiveItem(id)
      this.modalWindowType = ModalWindowType.PURCHASE_ITEM
    },
    async handleModalWindowConfirmation() {
      await this.purchaseItem(this.activeItem.id)
      this.discardActiveItem()
      this.modalWindowType = ModalWindowType.PURCHASE_RESULT
    },
    handleModalWindowClose(type) {
      this.modalWindowType = null
      switch (type) {
        case ModalWindowType.PURCHASE_ITEM:
          this.discardActiveItem()
          break
        case ModalWindowType.PURCHASE_RESULT:
          this.discardShoppingSuccess()
          break
      }
    }
  }
}
</script>

<style lang="scss" scoped>
  .shop {
    &__posessed-items {
      margin-top: 0;
      padding: 4px 8px;
      border-bottom: $border;
    }

    &__available-items {
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    &__available-items-list {
      margin: 0 64px 32px;
      max-width: 900px;
    }
  }
</style>
