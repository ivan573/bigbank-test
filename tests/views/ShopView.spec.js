import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VueRouter from 'vue-router'
import { ModalWindowType, Route } from '@/const'
import ShopView from '@/views/ShopView.vue'

describe('ShopView.vue', () => {
  const gameId = 'gameId'
  const items = [
    {
      cost: 50,
      id: 'hpot',
      name: 'Healing potion'
    },
    {
      cost: 100,
      id: 'tricks',
      name: 'Book of Tricks'
    },
    {
      cost: 300,
      id: 'mtrix',
      name: 'Book of Megatricks'
    }
  ]
  const activeItem = items[0]
  const posessedItems = []
  const shoppingSuccess = true

  const computed = {
    gameId() { return gameId },
    items() { return items },
    activeItem() { return activeItem },
    posessedItems() { return posessedItems },
    shoppingSuccess() { return shoppingSuccess }
  }

  it('matches snapshot', () => {
    const wrapper = mount(ShopView, { computed })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('whenever an item is clicked the component updates active item and displays purchase modal window', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const updateActiveItem = jest.fn()
    const shop = {
      namespaced: true,
      mutations: { updateActiveItem }
    }
    const store = new Vuex.Store({ modules: { shop } })

    const data = () => {
      return {
        modalWindowType: null
      }
    }

    const wrapper = mount(ShopView, {
      localVue,
      store,
      data,
      computed
    })

    const itemsElement = wrapper.find('.shop__available-items-list')

    itemsElement.vm.$emit('itemClick', items[0].id)

    expect(updateActiveItem).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.PURCHASE_ITEM)
  })

  it('whenever modal window emits confirm event component calls purchase action, discards active item and updates modal window type', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const purchaseItem = jest.fn()
    const discardActiveItem = jest.fn()
    const shop = {
      namespaced: true,
      mutations: { discardActiveItem },
      actions: { purchaseItem }
    }
    const store = new Vuex.Store({ modules: { shop } })

    const modalWindowType = ModalWindowType.PURCHASE_ITEM
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(ShopView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.shop__modal-window')

    await modalWindowElement.vm.$emit('confirm', modalWindowType)

    await wrapper.vm.$nextTick()

    expect(purchaseItem).toHaveBeenCalledTimes(1)
    expect(discardActiveItem).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.PURCHASE_RESULT)
  })

  it('whenever modal window emits confirm event component calls purchase action, discards active item and updates modal window type', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const purchaseItem = jest.fn()
    const discardActiveItem = jest.fn()
    const shop = {
      namespaced: true,
      mutations: { discardActiveItem },
      actions: { purchaseItem }
    }
    const store = new Vuex.Store({ modules: { shop } })

    const modalWindowType = ModalWindowType.PURCHASE_ITEM
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(ShopView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.shop__modal-window')

    await modalWindowElement.vm.$emit('confirm', modalWindowType)

    await wrapper.vm.$nextTick()

    expect(purchaseItem).toHaveBeenCalledTimes(1)
    expect(discardActiveItem).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.PURCHASE_RESULT)
  })

  it('whenever a purchase modal window emits close event component discards active item and modal window type', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardActiveItem = jest.fn()
    const shop = {
      namespaced: true,
      mutations: { discardActiveItem }
    }
    const store = new Vuex.Store({ modules: { shop } })

    const modalWindowType = ModalWindowType.PURCHASE_ITEM
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(ShopView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.shop__modal-window')

    await modalWindowElement.vm.$emit('close', modalWindowType)

    await wrapper.vm.$nextTick()

    expect(discardActiveItem).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.modalWindowType).toBe(null)
  })

  it('whenever a purchase result modal window emits close event component discards purchase result message and modal window type', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardShoppingSuccess = jest.fn()
    const shop = {
      namespaced: true,
      mutations: { discardShoppingSuccess }
    }
    const store = new Vuex.Store({ modules: { shop } })

    const modalWindowType = ModalWindowType.PURCHASE_RESULT
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(ShopView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.shop__modal-window')

    modalWindowElement.vm.$emit('close', modalWindowType)

    expect(discardShoppingSuccess).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.modalWindowType).toBe(null)
  })

  it('if the modal window has the purchase type component provides active item data to render', () => {
    const modalWindowType = ModalWindowType.PURCHASE_ITEM
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(ShopView, { data, computed })

    expect(wrapper.vm.modalWindowData).toEqual(wrapper.vm.activeItem)
  })

  it('if the modal window has the purchase result type and purchase was successful component provides corresponding message to render', () => {
    const modalWindowType = ModalWindowType.PURCHASE_RESULT
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(ShopView, { data, computed })

    expect(wrapper.vm.modalWindowData).toEqual({ message: 'You have purchased the item' })
  })

  it('if the modal window has the purchase result type and purchase was not successful component provides corresponding message to render', () => {
    const modalWindowType = ModalWindowType.PURCHASE_RESULT
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(ShopView, { data, computed: { ...computed, shoppingSuccess() { return false } } })

    expect(wrapper.vm.modalWindowData).toEqual({ message: 'You cannot purchase this item' })
  })

  it('redirects to root path if the game is not initialized', () => {
    const localVue = createLocalVue()
    localVue.use(VueRouter)

    const router = new VueRouter()
    router.push(Route.SHOP)

    const wrapper = shallowMount(ShopView, {
      localVue,
      router,
      computed: {
        ...computed,
        gameId() { return null }
      }
    })

    expect(wrapper.vm.$route.path).toBe(Route.HOME)
  })
})
