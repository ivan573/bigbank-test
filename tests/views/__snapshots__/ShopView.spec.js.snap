// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopView.vue matches snapshot 1`] = `
<div
  class="shop"
>
  <!---->
   
  <div
    class="shop__available-items"
  >
    <h2
      class="shop__available-items-title"
    >
      Available items
    </h2>
     
    <ul
      class="game-items shop__available-items-list"
    >
      <li
        class="game-items__item"
      >
        <p
          class="game-items__item-data"
        >
          Healing potion
        </p>
         
        <p
          class="game-items__item-data"
        >
          Price: 50
        </p>
      </li>
      <li
        class="game-items__item"
      >
        <p
          class="game-items__item-data"
        >
          Book of Tricks
        </p>
         
        <p
          class="game-items__item-data"
        >
          Price: 100
        </p>
      </li>
      <li
        class="game-items__item"
      >
        <p
          class="game-items__item-data"
        >
          Book of Megatricks
        </p>
         
        <p
          class="game-items__item-data"
        >
          Price: 300
        </p>
      </li>
    </ul>
     
    <!---->
  </div>
</div>
`;
